### PR TITLE
Allow ls works inside dir with [] brackets

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description
Fixes: #12429

To fix the issue, we need to pass the `input pattern` itself to `glob_from` function,  but currently on latest main, nushell pass `expanded path of input pattern` to `glob_from` function.
It causes globbing failed if expanded path includes `[]` brackets.

It's a pity that I have to duplicate `nu_engine::glob_from` function into `ls`, because `ls` might convert from `NuGlob::NotExpand` to `NuGlob::Expand`, in that case, `nu_engine::glob_from` won't work if user want to ls for a directory which includes tilde:
```
mkdir "~abc"
ls "~abc"
```
So I need to duplicate `glob_from` function and pass original `expand_tilde` information.

# User-Facing Changes
Nan

# Tests + Formatting
Done

# After Submitting
Nan

### CODE REVIEW

File: crates/nu-command/src/filesystem/ls.rs [hunk 1]
```diff
@@ -900,3 +890,114 @@ mod windows_helper {
...
+            if let Some(comp) = path.components().nth(c) {
+                just_pattern.push(comp);
+            }
+        }
+        if no_glob_for_pattern {
+            just_pattern = PathBuf::from(nu_glob::Pattern::escape(&just_pattern.to_string_lossy()));
+        }
+
+        // Now expand `p` to get full prefix
+        let path = expand_path_with(p, cwd, expand_tilde && pattern.item.is_expand());
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/ls.rs:943]: This is one of the only 2 difference lines between `nu_engine::glob_from` and this `glob_from`

File: crates/nu-command/src/filesystem/ls.rs [hunk 2]
```diff
@@ -900,3 +890,114 @@ mod windows_helper {
...
+        let escaped_prefix = PathBuf::from(nu_glob::Pattern::escape(&path.to_string_lossy()));
+
+        (Some(path), escaped_prefix.join(just_pattern))
+    } else {
+        let path = PathBuf::from(&pattern.item.as_ref());
+        let path = expand_path_with(path, cwd, expand_tilde && pattern.item.is_expand());
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/ls.rs:949]: This is the other difference line

**Comment by YizhePKU** [comment]
I think there're better ways to fix this. What do you think about [my comment](https://github.com/nushell/nushell/issues/12429#issuecomment-2071764413)?

**Comment by WindSoilder** [comment]
Thank you for the idea, I have tried to implement it, but it failed out testing.
It doesn't work for the following example:
```
mkdir abcd/* abcd/?
touch abcd/*/test.txt abcd/?/test.txt
ls abcd/*
```
If we use `read_dir`, it will only read `abcd/*`, but it should read both `abcd/*` and `abcd/?`

Here is my attempt: https://github.com/WindSoilder/nushell/pull/1

**Comment by YizhePKU** [comment]
It doesn't work because we're skipping the glob expansion when there's an exact match, even when the argument should be treated as a glob pattern.

Here's how I think things should work:

Step 1: Decide if the argument should be treated as a literal string or a glob pattern. The argument should be treated as a glob argument if it is unquoted **and** it contains no meta characters (we can use `Pattern::escape` to check this).

Step 2: If the path argument is treated as a literal string, check if it's actually a directory (or a symlink to a directory). If true, use `fs::read_dir()` and return its content; otherwise just return the path itself.

If the path argument is treated as a glob pattern, **perform the glob expansion unconditionally**, then return the result without further action.

---

This should fix the bug without changing the current behavior of `ls`, although I'm not sure if that behavior is the right one. I'll try it and see if it works.

**Comment by WindSoilder** [comment]
Thanks!  I think it's a good idea, just refactored it to use it

File: crates/nu-command/src/filesystem/ls.rs [hunk 3]
```diff
@@ -233,22 +234,20 @@ fn ls_for_one_pattern(
 
     // it indicates we need to append an extra '*' after pattern for listing given directory
     // Example: 'ls directory' -> 'ls directory/*'
```
Comment (YizhePKU) [inline crates/nu-command/src/filesystem/ls.rs:236]: Need to update comments here. We're no longer appending '*' to patterns.

File: crates/nu-engine/src/glob_from.rs [hunk 4]
```diff
@@ -6,7 +6,7 @@ use std::{
     path::{Component, Path, PathBuf},
 };
 
-const GLOB_CHARS: &[char] = &['*', '?', '['];
+pub const GLOB_CHARS: &[char] = &['*', '?', '['];
```
Comment (YizhePKU) [inline crates/nu-engine/src/glob_from.rs:9]: I think `GLOB_CHARS` is an implementation detail. It's probably better to duplicate this in your code than making it public.

Alternatively, maybe `s == Pattern::escape(s)` is a better way to check if `s` looks like a glob pattern?

File: crates/nu-command/src/filesystem/ls.rs [hunk 5]
```diff
@@ -900,3 +874,21 @@ mod windows_helper {
         false
     }
 }
+
+#[allow(clippy::type_complexity)]
+fn read_dir(
+    f: PathBuf,
+) -> Result<
+    (
+        Option<PathBuf>,
```
Comment (YizhePKU) [inline crates/nu-command/src/filesystem/ls.rs:883]: We don't have to follow the signature of `glob_from()`. I think just returning the iterator is cleaner.

**Comment by YizhePKU** [comment]
Looks good! I've left a few comments.

**Comment by WindSoilder** [comment]
I think it's ok to land this

**Pull request merged**